central/auth/service/service_impl_test.go

```diff
@@ -36,7 +36,7 @@ type authServiceAccessControlTestSuite struct {
 }
 
 func (s *authServiceAccessControlTestSuite) SetupSuite() {
-	s.svc = New()
+	s.svc = &serviceImpl{}
```

Comment:
[not about this line & nitpick]
I understand the idea that there's nothing interesting to test in M2M configs CRUD operations. All things are pretty standard and were tested over and over again.

However, as we establish a contract in `.proto` definition, there's a value to have regression tests that prevent the implementation to deviate from the declared contract.

My intent is not to ask you to write a bunch of boring code just to pass the pull request stage. I'm interested in what you think about balance between value in preventing regressions on one hand and cost of writing and maintaining extra tests on the other.

Reply:
Hmm, it's a good question.

I personally feel like the service layer itself in the current state doesn't do much w.r.t verifying the specific errors, but instead relies on the correct error (i.e. in this case `errox`) to be returned.

Having tests on those parts (like the validation as well as the datastore) and potentially within the postgres framework to ensure the correct `errox` is returned is enough for me without writing additional tests on the service layer. I feel like they'd be too much of a duplication from the datastore ones, and using mocks here defeats the purpose.

Reply:
> and using mocks here

That's **exactly** the problem to put a pin on. Testing with mocks (1) creates lots of testing code and (2) couples tests with the underlying implementation details. Mocks aren't needed if you move datastore tests to the service level (like I suggested in my other comment).

In my experience, there is no benefit of writing tests with mocks. Rather, there's a benefit to restructure code so that the aggregate behavior is tested.

ORMs and data access layers are a good example of that. Would you be interested to check that insertion inserts and selection selects, or would you be interested to check that certain service operation that uses insertion does modify the store? Rhetorical question because you already know how I think about it.

Reply:
So, since the tests have moved, is there anything in addition you would like to see?